_source/_posts/2021-10-27-k8s-api-server-oidc.md

```diff
@@ -0,0 +1,551 @@
+
+# Using Okta as an OIDC provider to secure the API server.
+
+> Okta, Inc. is an identity and access management company providing cloud software that helps companies manage and secure user authentication into applications and for developers to build identity controls into applications, websites, web services, and devices. Okta is a certified OpenID Connect provider.
+
+Let us see how we can secure the Kubernetes API server using Okta as an OIDC provider and use RBAC to control access right from the Okta admin console. If you are using Amazon EKS, then check [this](/blog/2021/10/08/secure-access-to-aws-eks) specific tutorial for using Okta OIDC with EKS.
+
+## What You'll Need to Get Started
+
+Before you start trying this out, make sure you have access to the following.
+
+- An Okta account. You can sign up for a free account [here](https://www.okta.com/free-trial/). If you like, you can use another OIDC provider or [Dex](https://github.com/dexidp/dex), and the steps should be similar.
+- A Kubernetes cluster. I'm using [k3d](https://k3d.io/) to run a local [k3s](https://k3s.io/) cluster. You can use any Kubernetes distribution, including managed PaaS like Amazon EKS, AKS, and GKE, and so on.
+- kubectl installed on your machine.
+- Terraform installed on your machine. This is not required if you do the Okta configuration via the [Okta admin console](https://login.okta.com/) GUI.
+
+## Setup an Okta OIDC application and authorization server
+
+You can achieve OIDC login for the cluster by creating a simple OIDC application with Okta either using the Okta CLI or the Admin console. But with an OIDC application alone, you would have to use the client secret to authenticate from kubectl or any other client library. Which does not scale and is not better than default k8s authentication mechanisms as you won't have granular controls over users and roles. For a more helpful setup, we would need an OIDC application and an authorization server with customized claims and policies for Kubernetes. This way, we can make use of Okta to manage users and permissions as well.
+
+There are multiple ways to setup an OIDC application and authorization server in Okta. If you prefer to do this via a GUI, then follow the **Configure Your Okta Org section** from [this article](/blog/2021/10/08/secure-access-to-aws-eks#configure-your-okta-org) to do it via the [Okta Admin console](https://login.okta.com/).
+
+In this tutorial, we will use Terraform to configure the Okta part so that you can reuse the code for any automation required. Let's dive into each step needed.
+
+### Setup Terraform
+
+If you don't have Terraform installed, follow [these instructions](https://learn.hashicorp.com/tutorials/terraform/install-cli) and set it up.
+
+You can find the complete Terraform source code for this article in this [GitHub repo](https://github.com/oktadev/k8s-okta-oidc-terraform)
```

Comment:
```suggestion
You can find the complete Terraform source code for this article in [this GitHub repo](https://github.com/oktadev/k8s-okta-oidc-terraform)
```

Reply:
I would suggest renaming the repo to following our conventions: `okta-keywords-example`. We use the `okta-` prefix and `-example` suffix to identify repos that are associated with blog posts.